### PR TITLE
Python UDF Support To Templates

### PR DESCRIPTION
**Pull request by fozzie15**
Add python UDF support to templates with existing JS support.

### CODE REVIEW

File: v2/pubsub-binary-to-bigquery/src/main/resources/requirements.txt
Comment (Polber) [inline v2/pubsub-binary-to-bigquery/src/main/resources/requirements.txt:1]: This is not needed

File: v2/googlecloud-to-googlecloud/src/test/resources/TextIOToBigQueryTest/pyudf.py [hunk 1]
```diff
@@ -0,0 +1,21 @@
+##
+# Copyright (C) 2018 Google LLC
```
Comment (Polber) [inline v2/googlecloud-to-googlecloud/src/test/resources/TextIOToBigQueryTest/pyudf.py:2]: nit: 2024

File: v2/googlecloud-to-googlecloud/src/test/java/com/google/cloud/teleport/v2/templates/PubSubToBigQueryIT.java [hunk 2]
```diff
@@ -285,4 +302,74 @@ private void basePubsubToBigQueryWithReload(
     assertThatBigQueryRecords(reloadedRecords)
         .hasRecordsUnordered(List.of(Map.of("id", 11, "job", testName, "name", "alower: message")));
   }
+
+  private void pubsubToBigQueryWithPythonUdf(
```
Comment (Polber) [inline v2/googlecloud-to-googlecloud/src/test/java/com/google/cloud/teleport/v2/templates/PubSubToBigQueryIT.java:306]: nit: have 1 IT with `paramsAdder`

**Review by Polber — CHANGES_REQUESTED** [review]

File: plugins/core-plugin/src/main/resources/Dockerfile-xlang-template [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+# Create Distroless xlang image compatible with YamlTemplate #
+#============================================================#
+FROM gcr.io/dataflow-templates-base/python311-template-launcher-base:latest
+
+
+ARG REQUIREMENTS_FILE=requirements.txt
+
+
+COPY --from=busybox:1.35.0-uclibc /bin/sh /bin/sh
+COPY --from=busybox:1.35.0-uclibc /bin/ls /bin/ls
```
[the remark this answers is not in the document]
Comment (fozzie15) [reply plugins/core-plugin/src/main/resources/Dockerfile-xlang-template:17]: Removed.

File: v2/common/src/main/java/com/google/cloud/teleport/v2/transforms/PythonExternalTextTransformer.java [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+    }
+
+    @Override
+    public PCollection<Row> expand(PCollection<Row> elements) {
+      return elements.apply(
+          PythonExternalTransform.<PCollection<Row>, PCollection<Row>>from("__constructor__")
+              .withArgs(
+                  PythonCallableSource.of(
+                      String.format(
+                          "import apache_beam as beam\n"
```
[the remark this answers is not in the document]
Comment (fozzie15) [reply v2/common/src/main/java/com/google/cloud/teleport/v2/transforms/PythonExternalTextTransformer.java:103]: Agreed

File: v2/googlecloud-to-googlecloud/src/main/java/com/google/cloud/teleport/v2/templates/PubSubToBigQuery.java [hunk 5]
```diff
@@ -424,4 +503,45 @@ public void processElement(ProcessContext context) {
...
+        LOG.info(String.valueOf(failsafeObj));
+        context.output(UDF_DEADLETTER_OUT, failsafeObj);
+      } else {
+        PubsubMessage transformedMessage =
+            rowToPubSubMessage(element.getValue("transformed_messsage"));
+        FailsafeElement failsafeObj =
+            FailsafeElement.of(
+                originalMessage,
+                new String(transformedMessage.getPayload(), StandardCharsets.UTF_8));
+        LOG.info(String.valueOf(failsafeObj));
```
[the remark this answers is not in the document]
Comment (fozzie15) [reply v2/googlecloud-to-googlecloud/src/main/java/com/google/cloud/teleport/v2/templates/PubSubToBigQuery.java:532]: Done.

File: v2/googlecloud-to-googlecloud/src/main/java/com/google/cloud/teleport/v2/templates/PubSubToBigQuery.java [hunk 6]
```diff
@@ -424,4 +503,45 @@ public void processElement(ProcessContext context) {
...
+
+    public PubsubMessage rowToPubSubMessage(Row row) {
+      assert row != null;
+      String messageId = row.getValue("messageId");
+      String payload = row.getValue("message");
+      Map<String, String> attributeMap = row.getValue("attributes");
+
+      assert payload != null;
+      return new PubsubMessage(payload.getBytes(), attributeMap, messageId);
+    }
```
[the remark this answers is not in the document]
Comment (fozzie15) [reply v2/googlecloud-to-googlecloud/src/main/java/com/google/cloud/teleport/v2/templates/PubSubToBigQuery.java:545]: Will do as part of future cleanup.

File: v2/googlecloud-to-googlecloud/src/test/resources/TextIOToBigQueryTest/pyudf.py (hunk 1 above)
Comment (fozzie15) [reply v2/googlecloud-to-googlecloud/src/test/resources/TextIOToBigQueryTest/pyudf.py:2]: Done.

File: v2/googlecloud-to-googlecloud/src/test/java/com/google/cloud/teleport/v2/templates/PubSubToBigQueryIT.java (hunk 2 above)
Comment (fozzie15) [reply v2/googlecloud-to-googlecloud/src/test/java/com/google/cloud/teleport/v2/templates/PubSubToBigQueryIT.java:306]: Will do as part of future cleanup.

File: v2/googlecloud-to-googlecloud/src/main/resources/requirements.txt [hunk 7]
```diff
@@ -0,0 +1 @@
+apache-beam[gcp,aws, azure, dataframe]==2.53.0
```
[the remark this answers is not in the document]
Comment (fozzie15) [reply v2/googlecloud-to-googlecloud/src/main/resources/requirements.txt:1]: Done.

File: v2/googlecloud-to-googlecloud/src/main/java/com/google/cloud/teleport/v2/templates/PubSubToBigQuery.java [hunk 8]
```diff
@@ -374,22 +411,64 @@ static class PubsubMessageToTableRow
...
-                  "InvokeUDF",
-                  FailsafeJavascriptUdf.<PubsubMessage>newBuilder()
-                      .setFileSystemPath(options.getJavascriptTextTransformGcsPath())
-                      .setFunctionName(options.getJavascriptTextTransformFunctionName())
-                      .setReloadIntervalMinutes(
-                          options.getJavascriptTextTransformReloadIntervalMinutes())
-                      .setSuccessTag(UDF_OUT)
-                      .setFailureTag(UDF_DEADLETTER_OUT)
-                      .build());
+      LOG.error("Do we even see this log?\n");
```
[the remark this answers is not in the document]
Comment (fozzie15) [reply v2/googlecloud-to-googlecloud/src/main/java/com/google/cloud/teleport/v2/templates/PubSubToBigQuery.java:414]: Done.

File: v2/googlecloud-to-googlecloud/src/main/java/com/google/cloud/teleport/v2/templates/PubSubToBigQuery.java [hunk 9]
```diff
@@ -374,22 +411,64 @@ static class PubsubMessageToTableRow
...
+                // Map the incoming messages into FailsafeElements so we can recover from failures
+                // across multiple transforms.
+                .apply(
+                    "MapToRecord",
+                    PythonExternalTextTransformer.FailsafeRowPythonExternalUdf
+                        .pubSubMappingFunction())
+                .setRowSchema(PythonExternalTextTransformer.FailsafeRowPythonExternalUdf.ROW_SCHEMA)
+                .setCoder(
+                    RowCoder.of(
+                        PythonExternalTextTransformer.FailsafeRowPythonExternalUdf.ROW_SCHEMA))
```
[the remark this answers is not in the document]
Comment (fozzie15) [reply v2/googlecloud-to-googlecloud/src/main/java/com/google/cloud/teleport/v2/templates/PubSubToBigQuery.java:436]: Will remove as part of future cleanup.

File: v2/pubsub-binary-to-bigquery/src/main/resources/requirements.txt
Comment (fozzie15) [reply v2/pubsub-binary-to-bigquery/src/main/resources/requirements.txt:1]: Leaving as future supported templates will use this

File: v2/googlecloud-to-googlecloud/src/main/java/com/google/cloud/teleport/v2/templates/FirestoreToBigQuery.java [hunk 10]
```diff
@@ -138,27 +164,67 @@ public static void main(String[] args) {
...
+                  .setGqlQuery(options.getFirestoreReadGqlQuery())
+                  .setProjectId(options.getFirestoreReadProjectId())
+                  .setNamespace(options.getFirestoreReadNamespace())
+                  .build())
+          .apply(
+              "MapToRecord",
+              PythonExternalTextTransformer.FailsafeRowPythonExternalUdf.stringMappingFunction())
+          .setRowSchema(PythonExternalTextTransformer.FailsafeRowPythonExternalUdf.ROW_SCHEMA)
+          .setCoder(
+              RowCoder.of(PythonExternalTextTransformer.FailsafeRowPythonExternalUdf.ROW_SCHEMA))
```
[the remark this answers is not in the document]
Comment (fozzie15) [reply v2/googlecloud-to-googlecloud/src/main/java/com/google/cloud/teleport/v2/templates/FirestoreToBigQuery.java:186]: Will change as part of future cleanup.

File: plugins/core-plugin/src/main/resources/Dockerfile-xlang-template [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+    && pip download --no-cache-dir --dest /tmp/dataflow-requirements-cache -r $REQUIREMENTS_FILE
+
+COPY --from=java-base /usr/local/jdk-11.0.20+8 /usr/local/jdk-11.0.20+8
+RUN ln -s /usr/local/jdk-11.0.20+8/bin/java /usr/local/bin/java
+COPY --from=java-base /opt/google/dataflow/java_template_launcher /opt/google/dataflow/java_template_launcher
+COPY --from=java-base /usr/licenses/LICENSE-java11 /usr/licenses/LICENSE-java11
+
+WORKDIR /template
+
+ENV PIP_NO_DEPS=True
```
[the remark this answers is not in the document]
Comment (fozzie15) [reply plugins/core-plugin/src/main/resources/Dockerfile-xlang-template:49]: Will change as part of future cleanup.

File: plugins/templates-maven-plugin/src/main/java/com/google/cloud/teleport/plugin/maven/TemplatesStageMojo.java [hunk 12]
```diff
@@ -338,18 +338,30 @@ protected String stageFlexTemplate(
...
     File imageSpecFile = generator.saveImageSpec(definition, imageSpec, outputClassesDirectory);
 
+    File xlangOutputDir;
+    File metadataFile;
+    File commandSpecFile;
+    if (definition.getTemplateAnnotation().type() == TemplateType.XLANG) {
+      xlangOutputDir =
+          new File(outputClassesDirectory.getPath() + "/" + containerName + "/resources");
+      metadataFile =
+          generator.saveMetadata(definition, imageSpec.getMetadata(), outputClassesDirectory);
```
[the remark this answers is not in the document]
Comment (fozzie15) [reply plugins/templates-maven-plugin/src/main/java/com/google/cloud/teleport/plugin/maven/TemplatesStageMojo.java:350]: Fixed.

File: v2/common/src/main/java/com/google/cloud/teleport/v2/transforms/PythonExternalTextTransformer.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (fozzie15) [reply v2/common/src/main/java/com/google/cloud/teleport/v2/transforms/PythonExternalTextTransformer.java:103]: Will be changed as part of future cleanup.

File: v2/common/src/main/java/com/google/cloud/teleport/v2/transforms/PythonExternalTextTransformer.java [hunk 13]
```diff
@@ -0,0 +1,211 @@
...
+    public static final Schema FAILSAFE_SCHEMA =
+        Schema.builder()
+            .addNullableRowField("original", ROW_SCHEMA)
+            .addNullableRowField("transformed", ROW_SCHEMA)
+            .addNullableStringField("error_message")
+            .addNullableStringField("stack_trace")
+            .build();
+
+    public static String pubsub = "pubsub";
+    public static String string = "string";
```
[the remark this answers is not in the document]
Comment (fozzie15) [reply v2/common/src/main/java/com/google/cloud/teleport/v2/transforms/PythonExternalTextTransformer.java:170]: Removed.

File: plugins/core-plugin/src/main/resources/Dockerfile-xlang-template (hunk 3 above)
[the remark this answers is not in the document]
Comment (Polber) [reply plugins/core-plugin/src/main/resources/Dockerfile-xlang-template:17]: Still showing up

[the remark this answers is not in the document]
Comment (fozzie15) [reply plugins/core-plugin/src/main/resources/Dockerfile-xlang-template]: Should be removed now. Hopefully gone.

**Review by Polber — DISMISSED** [review]

**Review by Polber — APPROVED** [review]

**Pull request merged**